bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/tag/TagResource.java

```diff
@@ -74,28 +84,195 @@ public class TagResource implements RESTResource {
+    public Response getTagAndSubTags(
+            @HeaderParam(HttpHeaders.ACCEPT_LANGUAGE) @Parameter(description = "language") @Nullable String language,
+            @PathParam("tagId") @Parameter(description = "tag id") String tagId) {
+        final Locale locale = localeService.getLocale(language);
+        String uid = tagId.trim();
+
+        SemanticTag tag = semanticTagRegistry.get(uid);
+        if (tag != null) {
+            List<EnrichedSemanticTagDTO> tagsDTO = semanticTagRegistry.getSubTree(tag).stream()
+                    .sorted((element1, element2) -> element1.getUID().compareTo(element2.getUID()))
+                    .map(t -> new EnrichedSemanticTagDTO(t.localized(locale), isEditable(t)))
+                    .collect(Collectors.toList());
+            return JSONResponse.createResponse(Status.OK, tagsDTO, null);
+        } else {
+            return getTagResponse(Status.NOT_FOUND, null, locale, "Tag " + uid + " does not exist!");
+        }
+    }
+
+    @POST
+    @RolesAllowed({ Role.ADMIN })
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Operation(operationId = "createSemanticTag", summary = "Creates a new semantic tag and adds it to the registry.", security = {
+            @SecurityRequirement(name = "oauth2", scopes = { "admin" }) }, responses = {
+                    @ApiResponse(responseCode = "201", description = "Created", content = @Content(schema = @Schema(implementation = EnrichedSemanticTagDTO.class))),
+                    @ApiResponse(responseCode = "400", description = "The tag identifier is invalid."),
+                    @ApiResponse(responseCode = "409", description = "A tag with the same identifier already exists.") })
+    public Response create(
+            @HeaderParam(HttpHeaders.ACCEPT_LANGUAGE) @Parameter(description = "language") @Nullable String language,
+            @Parameter(description = "tag data", required = true) EnrichedSemanticTagDTO data) {
+        final Locale locale = localeService.getLocale(language);
+
+        if (data.uid == null) {
+            return getTagResponse(Status.BAD_REQUEST, null, locale, "Tag identifier is required!");
+        }
+
+        String uid = data.uid.trim();
+
+        // check if a tag with this UID already exists
+        SemanticTag tag = semanticTagRegistry.get(uid);
+        if (tag != null) {
+            // report a conflict
+            return getTagResponse(Status.CONFLICT, tag, locale, "Tag " + uid + " already exists!");
+        }
+
+        // Extract the tag name and th eparent tag
+        // Check that the parent tag already exists
+        SemanticTag parentTag = null;
+        int lastSeparator = uid.lastIndexOf("_");
+        if (lastSeparator <= 0) {
+            return getTagResponse(Status.BAD_REQUEST, null, locale, "Invalid tag identifier " + uid);
+        }
+        String name = uid.substring(lastSeparator + 1);
+        parentTag = semanticTagRegistry.get(uid.substring(0, lastSeparator));
```

Comment:
I think the substring logic couuld be hidden in registry itself in `getParentTag(uid)` or similar method. It seems that this is some assumption of how UIDs should be structured which might actually span across multiple places.

Reply:
You're right, I will add such method.